### PR TITLE
StrikerVsGoalie and other self-play env improvements

### PR DESCRIPTION
**Pull request by andrewcoh**
### Proposed change(s)

This PR adds the StrikersVsGoalie env as well as some refactors to Tennis and Soccer

~~Tennis:~~
* ~~Removed the net causing a loss of a point on the serve. I think the agents were developing weird biases and I think it was partially the culprit behind the boring game play~~
* ~~Added a timestep penalty. The agents are now much more aggressive.~~

Reserving tennis changes for a separate PR

Soccer
* Extended the AgentSoccer.cs script to support strikers/goalies/generic soccer agents with different reward functions.
* Added a curriculum that should speed up learning 

Finally, I tweaked the hyperparameters for all these envs to be a bit more stable in training.

TODO:
* ~~Add Soccer and Tennis brains (currently retraining with curriculum/no net but training with the original envs yielded good results).~~
* ~~Add to docs for new StrikerVsGoalie env~~
* ~~Add to self-play docs regarding curriculum recommendation in specifying rewards~~

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [x] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [x] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs [hunk 1]
```diff
@@ -72,21 +104,21 @@ public void MoveAgent(float[] act)
...
 
         switch (rightAxis)
         {
             case 1:
-                dirToGo = transform.right * 0.3f;
+                dirToGo = transform.right * m_LateralSpeed;
                 break;
             case 2:
-                dirToGo = transform.right * -0.3f;
+                dirToGo = transform.right * -m_LateralSpeed;
```
Comment (surfnerd) [inline Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs:121]: noticed down below that `Time.deltaTime` is used for rotation but not for `AddForce`.  Have any ideas why?

File: ml-agents/mlagents/trainers/tests/test_simple_rl.py [hunk 2]
```diff
@@ -402,7 +402,7 @@ def test_simple_asymm_ghost_fails(use_discrete):
     processed_rewards = [
         default_reward_processor(rewards) for rewards in env.final_rewards.values()
     ]
-    success_threshold = 0.99
```
Comment (surfnerd) [inline ml-agents/mlagents/trainers/tests/test_simple_rl.py:405]: do we want to document why this changes over time?  Maybe a comment about why it was changed along with the date.

File: ml-agents/mlagents/trainers/ppo/trainer.py [hunk 3]
```diff
@@ -220,7 +220,9 @@ def _update_policy(self):
                 self._stats_reporter.add_stat(stat, val)
         self._clear_update_buffer()
 
-    def create_policy(self, brain_parameters: BrainParameters) -> TFPolicy:
+    def create_policy(
+        self, parsed_behavior_id: BehaviorIdentifiers, brain_parameters: BrainParameters
+    ) -> TFPolicy:
         """
         Creates a PPO policy to trainers list of policies.
         :param brain_parameters: specifications for policy construction
```
Comment (surfnerd) [inline ml-agents/mlagents/trainers/ppo/trainer.py:228]: add parsed_behavior_id doc.

**Comment by surfnerd** [comment]
code changes look ok to me.  Going to run the samples locally before approving

**Comment by ervteng** [comment]
Code looks good. Would add in the trained .nn to the PR when you have it

File: Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs [hunk 4]
```diff
@@ -107,8 +140,22 @@ public void MoveAgent(float[] act)
...
+        }
+        else if (position == Position.Striker)
+        {
+            // Existential penalty for Strikers
+            AddReward(-1f / 3000f);
+        }
+        else
+        {
+            // Existential penalty cumulant for Generic
+            timePenalty += -1f / 3000f;
```
Comment (andrewcoh) [inline Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs:157]: In this way, I accumulate timestep penalty and give it all at once at the end of the episode _if an agent wins_. The point here is so that there's never additional utility in 'losing faster' i.e. if each agent got timestep penalty irrespective of outcome but still utility in winning faster.  The trained soccertwos agents look much more stable defensively and I attribute that to this.

File: Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs [hunk 5]
```diff
@@ -18,11 +19,27 @@ public enum Team
...
+
     [HideInInspector]
     public Team team;
     float m_KickPower;
     int m_PlayerIndex;
     public SoccerFieldArea area;
+    float m_BallTouch;
+    public Position position;
+
+    float m_Power;
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs:37]:
```suggestion
    const float k_Power = 2000f;
```
I think this is better as a const.

File: Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs (hunk 4 above)
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs:157]: You only accumulate this time penalty for the "generic" position. Is that intended or were you trying to give it to the strikers ?

File: Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs [hunk 6]
```diff
@@ -107,8 +140,22 @@ public void MoveAgent(float[] act)
 
     public override void OnActionReceived(float[] vectorAction)
     {
-        // Existential penalty for strikers.
-        AddReward(-1f / 3000f);
+
+        if (position == Position.Goalie)
+        {
+            // Existential bonus for Goalies.
+            AddReward(1f / 3000f);
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs:147]: Make a const rather than copy `1f / 3000f` in these 3 places.

Comment (vincentpierre) [reply Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs:147]: Also, is this number dependent on the agent max_step ?

File: Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs [hunk 7]
```diff
@@ -147,9 +194,18 @@ public override void Heuristic(float[] actionsOut)
...
     {
-        var force = 2000f * m_KickPower;
+        var force = m_Power * m_KickPower;
+        if (position == Position.Goalie)
+        {
+            force = m_Power;
+        }
         if (c.gameObject.CompareTag("ball"))
         {
+            // Generic gets curriculum
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs:204]: What does this comment mean ?

File: Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs [hunk 8]
```diff
@@ -18,11 +19,27 @@ public enum Team
...
+        Goalie,
+        Generic
+    }
+
     [HideInInspector]
     public Team team;
     float m_KickPower;
     int m_PlayerIndex;
     public SoccerFieldArea area;
+    float m_BallTouch;
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs:34]: What does this number mean. Please add some comments or give more explicit names to these variables.

File: com.unity.ml-agents/CHANGELOG.md [hunk 9]
```diff
@@ -36,6 +36,7 @@ and this project adheres to
   communication between Unity and the Python process.
 - The obsolete `Agent` methods `GiveModel`, `Done`, `InitializeAgent`,
   `AgentAction` and `AgentReset` have been removed.
+- The GhostTrainer has been extended to support asymmetric games and the asymmetric example environment Strikers Vs. Goalie has been added.
```
Comment (vincentpierre) [inline com.unity.ml-agents/CHANGELOG.md:39]: use ` around GhostTrainer

File: docs/Learning-Environment-Examples.md [hunk 10]
```diff
@@ -354,35 +354,70 @@ you would like to contribute environments, please see our
...
-  - ball_scale: Specifies the scale of the ball in the 3 dimensions (equal
-    across the three dimensions)
-    - Default: 7.5
-    - Recommended minimum: 4
-    - Recommended maximum: 10
-  - gravity: Magnitude of the gravity
-    - Default: 9.81
-    - Recommended minimum: 6
-    - Recommended maximum: 20
+* Set-up: Environment where four agents compete in a 2 vs 2 toy soccer game.
```
Comment (vincentpierre) [inline docs/Learning-Environment-Examples.md:357]: Please be consistent with the rest of this document, it seems this one uses - instead of *

File: docs/Learning-Environment-Examples.md [hunk 11]
```diff
@@ -354,35 +354,70 @@ you would like to contribute environments, please see our
...
+* Goal:
+  * Get the ball into the opponent's goal while preventing
+  the ball from entering own goal.
```
Comment (vincentpierre) [inline docs/Learning-Environment-Examples.md:360]: unnecessary line split

File: docs/Learning-Environment-Examples.md [hunk 12]
```diff
@@ -354,35 +354,70 @@ you would like to contribute environments, please see our
...
+* Agents: The environment contains four agents, with the same
+  Behavior Parameters : SoccerTwos.
+* Agent Reward Function (dependent):
+    * +1 When ball enters opponent's goal.
```
Comment (vincentpierre) [inline docs/Learning-Environment-Examples.md:364]: No accumulated time penalty ?

File: docs/Learning-Environment-Examples.md [hunk 13]
```diff
@@ -354,35 +354,70 @@ you would like to contribute environments, please see our
...
-    - Default: 9.81
-    - Recommended minimum: 6
-    - Recommended maximum: 20
+* Set-up: Environment where four agents compete in a 2 vs 2 toy soccer game.
+* Goal:
+  * Get the ball into the opponent's goal while preventing
+  the ball from entering own goal.
+* Agents: The environment contains four agents, with the same
+  Behavior Parameters : SoccerTwos.
+* Agent Reward Function (dependent):
```
Comment (vincentpierre) [inline docs/Learning-Environment-Examples.md:363]: What does dependent mean ?

File: Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs [hunk 14]
```diff
@@ -18,11 +19,27 @@ public enum Team
         Purple = 1
     }
 
+    public enum Position
+    {
+        Striker,
+        Goalie,
+        Generic
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs:26]: IS this Position used in a scene?

File: Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs (hunk 4 above)
Comment (andrewcoh) [reply Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs:157]: That was intended. The strikers deserve it every timestep

File: Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs (hunk 14 above)
Comment (andrewcoh) [reply Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs:26]: It's a dropdown on the soccer agent script

File: docs/Learning-Environment-Examples.md (hunk 10 above)
Comment (andrewcoh) [reply docs/Learning-Environment-Examples.md:357]: I am shocked that I did that....

File: docs/Learning-Environment-Examples.md (hunk 13 above)
Comment (andrewcoh) [reply docs/Learning-Environment-Examples.md:363]: The agents share the reward function.

File: Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs (hunk 7 above)
Comment (andrewcoh) [reply Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs:204]: That only the generic agent gets curriculum. I can remove the if statement because the default value of m_BallTouch will be zero since it's set using curriculum.

File: Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs (hunk 6 above)
Comment (andrewcoh) [reply Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs]: I added a private float that's set in Initialize using maxStep.

File: Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs (hunk 14 above)
Comment (vincentpierre) [reply Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs:26]: But is it selected in one of the scenes? Are we testing that this setting actually learns ?

Comment (vincentpierre) [reply Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs:26]: (it is, in the 2v2)
